Support for describing images using CaptionBot

jcsteh opened the pull request:
### Link to issue number:
None.

### Summary of the issue:
Ideally, images would all have useful alternative text, but this is all too often not the case. This PR adds a command which uses Microsoft's [CaptionBot](https://www.captionbot.ai/) service to describe images.

### Description of how this pull request fixes the issue:
You press NVDA+alt+d to describe the image at the current navigator object. This uploads the image to CaptionBot and retrieves the description.

### Testing performed:
Tried on the NV Access logo on the NV Access website, the picture in the top table of the [Wikipedia article about labrador retrievers](https://en.wikipedia.org/wiki/Labrador_Retriever) and the Desktop.

### Known issues with pull request:
The CaptionBot API is not documented anywhere I could find. However, there are [several](https://pypi.python.org/pypi/captionbot) [publicly](https://www.npmjs.com/package/captionbot) [available](https://daniil.it/captionbot-clients/) [libraries](https://github.com/DeanF/imgcaption) which implement it. I also couldn't find any terms which specifically forbid its usage.

### Change log entry:
In New Features:

```
- NVDA can now use Microsoft's CaptionBot service to describe images. To describe the image at the current navigator object, press NVDA+alt+d. (#7476)
```

beqabeqa473 commented:
can we use wx.bitmap for image saving instead of pillow?

On 8/7/17, James Teh <notifications@github.com> wrote:
> ### Link to issue number:
> None.
>
> ### Summary of the issue:
> Ideally, images would all have useful alternative text, but this is all too
> often not the case. This PR adds a command which uses Microsoft's
> [CaptionBot] (https://www.captionbot.ai/) service to describe images.
>
> ### Description of how this pull request fixes the issue:
> You press NVDA+alt+c to describe the image at the current navigator object.
> This uploads the image to CaptionBot and retrieves the description.
>
> In order to write the image as a jpeg, NVDA now depends on Pillow, which has
> been added as a submodule.
>
> ### Testing performed:
> Tried on the NV Access logo on the NV Access website, the picture in the top
> table of the [Wikipedia article about labrador
> retrievers](https://en.wikipedia.org/wiki/Labrador_Retriever) and the
> Desktop.
>
> ### Known issues with pull request:
> 1. This adds Pillow as a dependency so we can write jpeg files. However, it
> includes a lot of stuff we don't need and is about 1 mb compressed. The
> question is whether this is an unacceptable increase in download size. Note
> that we will need this for any other cloud recognition services we might
> want to use.
> 2. The CaptionBot API is not documented anywhere I could find. However,
> there are [several](https://pypi.python.org/pypi/captionbot)
> [publicly](https://www.npmjs.com/package/captionbot)
> [available](https://daniil.it/captionbot-clients/)
> [libraries](https://github.com/DeanF/imgcaption) which implement it. I also
> couldn't find any terms which specifically forbid its usage.
>
> ### Change log entry:
> In New Features:
>
> ```
> - NVDA can now use Microsoft's CaptionBot service to describe images. To
> describe the image at the current navigator object, press NVDA+alt+c.
> (#7476)
> ```
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/nvaccess/nvda/pull/7476
>
> -- Commit Summary --
>
>   * Support for describing images using CaptionBot.
>
> -- File Changes --
>
>     M .gitmodules (3)
>     M copying.txt (12)
>     A include/pillow (1)
>     M readme.md (1)
>     A source/contentRecog/cloudRecog.py (96)
>     A source/extensionPoints.py (140)
>     M source/globalCommands.py (8)
>     A source/httpRequest.py (83)
>     M source/sourceEnv.py (1)
>     M user_docs/en/userGuide.t2t (10)
>
> -- Patch Links --
>
> https://github.com/nvaccess/nvda/pull/7476.patch
> https://github.com/nvaccess/nvda/pull/7476.diff
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/7476
>


-- 
with best regards beqa

josephsl commented:
Hi, may I request change of this command to account for the fact that in Word, it announces text comments? Thanks.

 

From: James Teh [mailto:notifications@github.com] 
Sent: Sunday, August 6, 2017 10:58 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: [nvaccess/nvda] Support for describing images using CaptionBot (#7476)

 


Link to issue number:


None.


Summary of the issue:


Ideally, images would all have useful alternative text, but this is all too often not the case. This PR adds a command which uses Microsoft's [CaptionBot] (https://www.captionbot.ai/) service to describe images.


Description of how this pull request fixes the issue:


You press NVDA+alt+c to describe the image at the current navigator object. This uploads the image to CaptionBot and retrieves the description.

In order to write the image as a jpeg, NVDA now depends on Pillow, which has been added as a submodule.


Testing performed:


Tried on the NV Access logo on the NV Access website, the picture in the top table of the Wikipedia article about labrador retrievers <https://en.wikipedia.org/wiki/Labrador_Retriever>  and the Desktop.


Known issues with pull request:


1.	This adds Pillow as a dependency so we can write jpeg files. However, it includes a lot of stuff we don't need and is about 1 mb compressed. The question is whether this is an unacceptable increase in download size. Note that we will need this for any other cloud recognition services we might want to use.
2.	The CaptionBot API is not documented anywhere I could find. However, there are several <https://pypi.python.org/pypi/captionbot>  publicly <https://www.npmjs.com/package/captionbot>  available <https://daniil.it/captionbot-clients/>  libraries <https://github.com/DeanF/imgcaption>  which implement it. I also couldn't find any terms which specifically forbid its usage.


Change log entry:


In New Features:

- NVDA can now use Microsoft's CaptionBot service to describe images. To describe the image at the current navigator object, press NVDA+alt+c. (#7476)
  _____  


You can view, comment on, or merge this pull request online at:


  https://github.com/nvaccess/nvda/pull/7476


Commit Summary


*	Support for describing images using CaptionBot.


File Changes


*	M .gitmodules <https://github.com/nvaccess/nvda/pull/7476/files#diff-0>  (3) 
*	M copying.txt <https://github.com/nvaccess/nvda/pull/7476/files#diff-1>  (12) 
*	A include/pillow <https://github.com/nvaccess/nvda/pull/7476/files#diff-2>  (1) 
*	M readme.md <https://github.com/nvaccess/nvda/pull/7476/files#diff-3>  (1) 
*	A source/contentRecog/cloudRecog.py <https://github.com/nvaccess/nvda/pull/7476/files#diff-4>  (96) 
*	A source/extensionPoints.py <https://github.com/nvaccess/nvda/pull/7476/files#diff-5>  (140) 
*	M source/globalCommands.py <https://github.com/nvaccess/nvda/pull/7476/files#diff-6>  (8) 
*	A source/httpRequest.py <https://github.com/nvaccess/nvda/pull/7476/files#diff-7>  (83) 
*	M source/sourceEnv.py <https://github.com/nvaccess/nvda/pull/7476/files#diff-8>  (1) 
*	M user_docs/en/userGuide.t2t <https://github.com/nvaccess/nvda/pull/7476/files#diff-9>  (10) 


Patch Links:


*	https://github.com/nvaccess/nvda/pull/7476.patch
*	https://github.com/nvaccess/nvda/pull/7476.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/7476> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkCfXHAYvIdBBNxYGsYWApD5hK4z1ks5sVqdegaJpZM4OvBcr> .

beqabeqa473 commented:
obj = api.getNavigatorObject()
		if obj.role != controlTypes.ROLE_GRAPHIC:
			ui.message(_('not an image!'))
			return
		x, y, width, height = obj.location
		image = wx.EmptyBitmap(width, height)
		mem = wx.MemoryDC(image)
		mem.Blit(0, 0, width, height, wx.ScreenDC(), x, y)
		image.SaveFile(self.imagePath, wx.BITMAP_TYPE_PNG)

jcsteh committed:
Support for describing images using CaptionBot.

This uses Microsoft's CaptionBot (https://www.captionbot.ai/) service to describe images.
You press NVDA+alt+d to describe the image at the current navigator object.

jcsteh commented:
[**@beqabeqa473**](https://github.com/beqabeqa473) commented on [7 Aug 2017, 16:06 GMT+10](https://github.com/nvaccess/nvda/pull/7476#issuecomment-320576653 "2017-08-07T06:06:36Z - Replied by Github Reply Comments"):
> can we use wx.bitmap for image saving instead of pillow?

Oh. I ... didn't even consider that possibility. I've now done that and we thus avoid the Pillow dependency. Thanks heaps for the idea!

[**@josephsl**](https://github.com/josephsl) commented on [7 Aug 2017, 16:09 GMT+10](https://github.com/nvaccess/nvda/pull/7476#issuecomment-320576995 "2017-08-07T06:09:06Z - Replied by Github Reply Comments"):
> Hi, may I request change of this command to account for the fact that in Word, it announces text comments?

Oh bleh. I missed this. I changed it to NVDA+alt+d (d for describe). That's the best I could come up with. :)

jcsteh commented:
@feerrenrut, I force pushed these changes into the HEAD commit, since you haven't started review and it's less confusing this way. I created a captionBotWithPillow branch in case we want to switch back to Pillow or for reference if we need Pillow for something else, but I doubt it.

feerrenrut reviewed (commented):
I tested this on a handful of images, overall the results seem pretty good. One concern I have is what will happen if someone exceeds the limit on the number of requests, assuming there is a limit.

I haven't finished inspecting the code closely yet, will go through it more carefully tomorrow.

derekriemer commented:
@jcsteh did you get rid of pillow for the ocr stuff then too while we're at it?

jcsteh commented:
UWP OCR never needed Pillow. It takes raw image data (don't need to convert to jpg) and we already have other code to do screen captures. I was only using Pillow to do jpg conversion.

derekriemer commented:
Go to the first graphic on that wikipedia article. Describe the image.

ERROR - scriptHandler.executeScript (03:14:01.536):
error executing script: <bound method GlobalCommands.script_recognizeWithCaptionBot of <globalCommands.GlobalCommands object at 0x0595E2D0>> with gesture u'alt+NVDA+d'
Traceback (most recent call last):
  File "scriptHandler.py", line 187, in executeScript
    script(gesture)
  File "globalCommands.py", line 2064, in script_recognizeWithCaptionBot
    recogUi.recognizeNavigatorObject(recog)
  File "contentRecog\recogUi.py", line 129, in recognizeNavigatorObject
    left, top, width, height = nav.location
TypeError: 'NoneType' object is not iterable

jcsteh commented:
[**@derekriemer**](https://github.com/derekriemer) commented on [4 Sep 2017, 19:14 GMT+10](https://github.com/nvaccess/nvda/pull/7476#issuecomment-326910592 "2017-09-04T09:14:59Z - Replied by Github Reply Comments"):
> TypeError: 'NoneType' object is not iterable

This happened because the object's location property returned None, which in turn probably happened because the object died (or your navigator object was locked somewhere else). Nevertheless, we should fail gracefully in this case. I've fixed this to report "Content not visible" in #7567.

feerrenrut commented:
We were not able to ensure we have permission to use this API, and as such, I'll close this issue for now.

leonardder commented:
While this pull request is closed, might it make sense to create a separate pr that only covers the CloudRecognizer base class in other for add-ons to extend on this easily?

michaelDCurran commented:
If you have a specific add-on in mind, then yes.

derekriemer commented:
We have prototypes. Unfortunately these cloud recognize services cost money and using one API key is not a smart idea. I could do a subscription service if necessary, or something.

leonardder commented:
I was thinking about https://github.com/ctoth/cloudocr , for example
